src/interrupt.cpp

```diff
@@ -0,0 +1,152 @@
+// (pollForInterrupts) always calls requestPolling to keep polling alive. The
+// background thread periodically attmeps to "collect" this request and if
+// successful re-schedules the pollForInterrupts function using
+// Py_AddPendingCall. This allows us to prevent the background thread from
+// continually scheduling pollForInterrupts even when the Python interpreter is
+// not running (because once pollForInterrupts is no longer being called by the
+// Python interpreter no additonal calls to pollForInterrupts will be
+// scheduled)
+class InterruptPollingSignal {
+public:
+  InterruptPollingSignal() : pollingRequested_(true) {}
+  
+  void requestPolling() {
+    lock_guard<mutex> lock(mutex_);
+    pollingRequested_ = true;
+  }
+  
+  bool collectRequest() {
+    lock_guard<mutex> lock(mutex_);
+    bool requested = pollingRequested_;
+    pollingRequested_ = false;
+    return requested;
+  }
+ 
+private:
+  InterruptPollingSignal(const InterruptPollingSignal& other); 
+  InterruptPollingSignal& operator=(const InterruptPollingSignal&);
+private:
+  mutex mutex_; 
+  bool pollingRequested_;
+};
+
+InterruptPollingSignal s_pollingSignal;
+
+extern "C" {
+
+// Forward declarations
+int pollForInterrupts(void*);
+void checkUserInterrupt(void*);
+  
+// Background thread which re-schedules pollForInterrupts on the main Python
+// interpreter thread every 100ms so long as the Python interpeter is still
+// running (when it stops running it will stop calling pollForInterrupts and
+// the polling signal will not be set).
+void interruptPollingWorker(void *) {
+  while(true) {
+    
+    // Throttle via sleep
+    this_thread::sleep_for(chrono::milliseconds(100));
+    
+    // Schedule polling on the main thread if the interpeter is still running
+    // Note that Py_AddPendingCall is documented to be callable from a background
+    // thread: "This function doesn’t need a current thread state to run, and it 
+    // doesn’t need the global interpreter lock."
+    // (see: https://docs.python.org/3/c-api/init.html#c.Py_AddPendingCall)
+    if (s_pollingSignal.collectRequest())
+      Py_AddPendingCall(pollForInterrupts, NULL);
+    
+  }
+}
+
+// Callback function scheduled to run on the main Python interpreter loop. This
+// is scheduled using Py_AddPendingCall, which ensures that it is run on the
+// main thread while the interpreter is executing. Note that we can't just have
+// this function keep reschedulding itself or the interpreter would be swamped
+// with just calling and re-calling this function. Rather, we need to throttle
+// the scheduling of the function by using a background thread + a sleep timer.
+int pollForInterrupts(void*) {
+  
+  // Check whether an interrupt has been requested by the user. If one
+  // has then set the Python interrupt flag (which will soon after result
+  // in a KeyboardInterrupt error being thrown).
+  if (R_ToplevelExec(checkUserInterrupt, NULL) == FALSE)
```

Comment:
Perhaps add in the comments that `R_ToplevelExec` returns `FALSE` if the execution of the associated function resulted in a jump to the top level? (and, when checking for interrupts, that will trigger and catch such a longjmp if the user has indeed requested an interrupt)